example.py

```diff
@@ -43,10 +44,16 @@ def device_updated(device):
 
 
 async def test_many():
-    device_options = [
-        aioshelly.ConnectionOptions("192.168.1.165", "admin", "test-password"),
-        aioshelly.ConnectionOptions("192.168.1.168"),
-    ]
+    device_options = []
+    with open("devices.json") as fp:
+        for line in fp:
+            obj = json.loads(line)
+            device_data = []
+            if "username" in obj:
+                device_data = [obj["ip"], obj["username"], obj["password"]]
+            else:
+                device_data = [obj["ip"]]
+            device_options.append(aioshelly.ConnectionOptions(device_data[0]))
```

Comment:
You're not using user/pass defined in `device_data`. There is also no reason to define `device_data` as intermediary. You can just read from `obj`.